[CI] Change the way tests are triggered with dynamo and inductor

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94539

Summary: Currently running PyTorch tests with dynamo and inductor is
controlled by environment variables, and CI sets them based on test
config name matching. Change them to use options of run_test.py.

desertfire committed:
Change the way tests are triggered with dynamo and inductor

Summary: Currently running PyTorch tests with dynamo and inductor is
controlled by an environment variable, which is less intuitive than
as options to run_test.py.

[ghstack-poisoned]

huydhn commented on torch/testing/_internal/common_utils.py:
```diff
@@ -2134,14 +2137,13 @@ def _run_with_retry(self, result=None, num_runs_left=0, report_only=True, num_re
             # torch._dynamo.reset()
+        elif TEST_WITH_TORCHINDUCTOR:
+            super_run = torch._dynamo.optimize("inductor")(super().run)
+            super_run(result=result)
```
Nit: Move `super_run(result=result)` out of the if-else as it runs for all branches anyway?

huydhn reviewed (approved):
LGTM! Only some tiny nits

huydhn commented on test/run_test.py:
```diff
@@ -1314,6 +1324,12 @@ def main():
+    if options.dynamo:
+        torch.testing._internal.common_utils.TEST_WITH_TORCHDYNAMO = True
+
+    if options.inductor:
```
Nit:  Does these two options mutually exclusive (only one can be set)?  If yes, you could add https://docs.python.org/3/library/argparse.html#mutual-exclusion to prevent people accidentally setting them both

huydhn commented on test/run_test.py:
```diff
@@ -1314,6 +1324,12 @@ def main():
+        torch.testing._internal.common_utils.TEST_WITH_TORCHDYNAMO = True
+
+    if options.inductor:
+        torch.testing._internal.common_utils.TEST_WITH_TORCHINDUCTOR = True
```
Ohh, good catch.  The argument would need to be passed to the child processes, i.e. like `--import-slow-tests`. +1

desertfire committed:
Update on "Change the way tests are triggered with dynamo and inductor"


Summary: Currently running PyTorch tests with dynamo and inductor is
controlled by an environment variable, which is less intuitive than
as options to run_test.py.

[ghstack-poisoned]

desertfire committed:
Update on "Change the way tests are triggered with dynamo and inductor"


Summary: Currently running PyTorch tests with dynamo and inductor is
controlled by environment variables, and CI sets them based on test
config name matching. Change them to use options of run_test.py.

[ghstack-poisoned]

desertfire commented:
@pytorchbot rebase

desertfire committed:
Update on "Change the way tests are triggered with dynamo and inductor"


Summary: Currently running PyTorch tests with dynamo and inductor is
controlled by environment variables, and CI sets them based on test
config name matching. Change them to use options of run_test.py.

[ghstack-poisoned]

desertfire commented:
@pytorchbot rebase

desertfire committed:
Update on "Change the way tests are triggered with dynamo and inductor"


Summary: Currently running PyTorch tests with dynamo and inductor is
controlled by environment variables, and CI sets them based on test
config name matching. Change them to use options of run_test.py.

[ghstack-poisoned]

huydhn reviewed (approved):
LGTM!

desertfire commented:
@pytorchbot merge

desertfire commented:
@pytorchbot merge -f "macos CI failure is irrelevant"
